datumaro/components/extractor.py

```diff
@@ -517,9 +581,40 @@ class DatasetItem:
         type=str, validator=not_empty)
     annotations = attrib(factory=list, validator=default_if_none(list))
     subset = attrib(converter=lambda v: v or DEFAULT_SUBSET_NAME, default=None)
+
+    # Currently unused
     path = attrib(factory=list, validator=default_if_none(list))
 
+    # TODO: introduce "media" field with type info. Replace image and pcd.
     image = attrib(type=Image, default=None)
+    pcd = attrib(type=bytes, default=None)
+    related_images = attrib(type=list, default=None)
+
+    def __attrs_post_init__(self):
+        assert not (self.has_image and self.has_pcd), \
+            "Can't set both image and point cloud info"
+
+    @related_images.validator
+    def _related_image_validator(self, attribute, related_images):
+        self.related_images = []
+
+        if not related_images:
+            return
+        for image in related_images:
+            if callable(image) or isinstance(image, np.ndarray):
+                image = Image(data=image)
+            elif isinstance(image, str):
+                image = Image(path=image)
+            assert isinstance(image, Image), type(image)
+
+            self.related_images.append(image)
+
+    @pcd.validator
+    def _pcd_validator(self, attribute, pcd):
+        if pcd:
+            assert isinstance(pcd, (bytes, str)) or callable(pcd), type(pcd)
+            self.pcd = pcd
```

Comment:
I don't think a validator is supposed to set the attribute itself (wouldn't that cause a recursive loop anyway?). You just need to check the type.

Reply:
Validators and converters only work on initialization. Refactored.